### PR TITLE
nrf_security: Enable CryptoCell as default on TF-M with code size improvements

### PR DESCRIPTION
**Pull request by frkv**
This PR does the following:

-Enables fine-grained control of psa_crypto_driver_wrappers enabled features making
 use of driver-specific support-configurations (prompt-less)
-Fixes issue with SHA-1 and AES GCM
-Cleanup of psa_crypto_driver_wrapper.c

ref: NCSDK-16500
ref: NCSDK-16515
ref: NCSDK-16702

Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>

### CODE REVIEW

**Review by tejlmand — APPROVED** [review]
lgtm

File: nrf_security/cmake/psa_crypto_config.cmake [hunk 1]
```diff
@@ -97,7 +105,6 @@ kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_ALG_RSA_PSS_CC3XX)
 kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_ALG_SHA_1_CC3XX)
 kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_ALG_SHA_224_CC3XX)
 kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_ALG_SHA_256_CC3XX)
-kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_ALG_SHA_384_CC3XX)
```
Comment (joerchan) [inline nrf_security/cmake/psa_crypto_config.cmake:100]: Commit messages does not mention why this was removed.

Is it not in use?

File: nrf_security/src/drivers/nrf_oberon/Kconfig [hunk 2]
```diff
@@ -63,7 +102,12 @@ config PSA_CRYPTO_DRIVER_ALG_ECDSA_OBERON
 	default y if !PSA_DEFAULT_OFF && PSA_WANT_ALG_ECDSA && \
 		     !PSA_CRYPTO_DRIVER_ALG_ECDSA_CC3XX
 
-# PSA_CRYPTO_DRIVER_ALG_GCM_OBERON - Currently not supported
+config PSA_CRYPTO_DRIVER_ALG_GCM_OBERON
+	bool
+	prompt "PSA AES GCM support - OBERON" if !PSA_PROMPTLESS
+	default y if !PSA_DEFAULT_OFF
+	depends on PSA_WANT_ALG_GCM && \
+		   !PSA_CRYPTO_DRIVER_ALG_GCM_CC3XX
```
Comment (joerchan) [inline nrf_security/src/drivers/nrf_oberon/Kconfig:110]: Oberon depends on CC3XX here?

Should this be PSA_CRYPTO_DRIVER_ALG_GCM_OBERON?

File: nrf_security/cmake/psa_crypto_config.cmake [hunk 3]
```diff
@@ -147,7 +147,11 @@ kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_ECC_SECP_R1_224_OBERON)
 kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_ECC_SECP_R1_256_OBERON)
 
 # Convert builtin driver configurations
-kconfig_check_and_set_base_to_one(MBEDTLS_PSA_BUILTIN_CC3XX)
```
Comment (joerchan) [inline nrf_security/cmake/psa_crypto_config.cmake:138]: Commit message does not mention why this was removed.
Is it correct?

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 4]
```diff
@@ -186,60 +155,39 @@ psa_status_t psa_driver_wrapper_sign_message(
...
                                       key_buffer_size,
                                       alg,
                                       input,
                                       input_length,
                                       signature,
                                       signature_size,
-                                      signature_length ) );
+                                      signature_length );
+    if( status != PSA_ERROR_NOT_SUPPORTED )
+        return( status );
```
Comment (mswarowsky) [inline nrf_security/src/psa_crypto_driver_wrappers.c:179]: Is there a system when `return( status )`  and `return status` is used

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 5]
```diff
@@ -2751,34 +2201,23 @@ psa_status_t psa_driver_wrapper_mac_compute(
...
+#if defined(PSA_CRYPTO_DRIVER_HAS_MAC_SUPPORT_CC3XX)
             status = cc3xx_mac_compute(attributes, key_buffer, key_buffer_size, alg,
                 input, input_length,
                 mac, mac_size, mac_length);
             /* Declared with fallback == true */
             if( status != PSA_ERROR_NOT_SUPPORTED )
                 return( status );
-#endif /* PSA_CRYPTO_DRIVER_CC3XX */
+#endif /* PSA_CRYPTO_DRIVER_HAS_MAC_SUPPORT_CC3XX */
 // TODO: Either remove or adjust when mac is available in Oberon
```
Comment (mswarowsky) [inline nrf_security/src/psa_crypto_driver_wrappers.c:2212]: Not directly linked with this PR shouldn't a TODO always linked to a Jira ticket so it is findable

File: nrf_security/src/drivers/nrf_oberon/Kconfig (hunk 2 above)
Comment (mswarowsky) [reply nrf_security/src/drivers/nrf_oberon/Kconfig:110]: It is supposed to either be CC3XX or Oberon, So Oberon is only used it the CC3XX support is  not used

File: nrf_security/cmake/psa_crypto_config.cmake (hunk 1 above)
Comment (frkv) [reply nrf_security/cmake/psa_crypto_config.cmake:100]: Fixed

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 6]
```diff
@@ -407,33 +334,33 @@ psa_status_t psa_driver_wrapper_sign_hash(
...
+            status = psa_sign_hash_builtin( attributes,
+                                            key_buffer,
+                                            key_buffer_size,
+                                            alg,
+                                            hash,
+                                            hash_length,
+                                            signature,
+                                            signature_size,
+                                            signature_length );
+            if( status != PSA_ERROR_NOT_SUPPORTED )
```
[the remark this answers is not in the document]
Comment (frkv) [reply nrf_security/src/psa_crypto_driver_wrappers.c:350]: Don't want to change with upstream psa_crypto_driver_wrappers.c. Won't fix

File: nrf_security/src/drivers/nrf_oberon/Kconfig (hunk 2 above)
Comment (frkv) [reply nrf_security/src/drivers/nrf_oberon/Kconfig:116]: nrf_oberon is only the fallback if CryptoCell is enabled

File: nrf_security/src/psa_crypto_driver_wrappers.c (hunk 4 above)
Comment (frkv) [reply nrf_security/src/psa_crypto_driver_wrappers.c:179]: Only in the upstream code for psa_crypto_driver_wrappers.c (the one that is copied here)

File: nrf_security/src/psa_crypto_driver_wrappers.c (hunk 5 above)
Comment (frkv) [reply nrf_security/src/psa_crypto_driver_wrappers.c]: Todos have been removed. Fixed!

File: nrf_security/src/drivers/builtin/Kconfig [hunk 7]
```diff
@@ -5,6 +5,41 @@
...
+	bool
+	default y
+	depends on MBEDTLS_PSA_BUILTIN_ALG_RSA_OAEP 		|| \
+		   MBEDTLS_PSA_BUILTIN_ALG_RSA_PKCS1V15_CRYPT
+
+config MBEDTLS_PSA_BUILTIN_HAS_ASYM_SIGN_SUPPORT
+	bool
+	default y
+	depends on MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA	|| \
+		   MBEDTLS_PSA_BUILTIN_ALG_ECDSA 			|| \
```
[the remark this answers is not in the document]
Comment (frkv) [reply nrf_security/src/drivers/builtin/Kconfig]: Tabs are used for myriads of other KConfigs in Zephyr. From what I can see this is the default in Zephyr
Won't fix here

File: nrf_security/cmake/psa_crypto_config.cmake (hunk 3 above)
Comment (frkv) [reply nrf_security/cmake/psa_crypto_config.cmake:138]: Fixed

File: nrf_security/src/drivers/nrf_cc3xx/Kconfig [hunk 8]
```diff
@@ -84,7 +142,7 @@ config PSA_CRYPTO_DRIVER_ALG_GCM_CC3XX
 	bool
 	prompt "PSA AES GCM support - cc3xx" if !PSA_PROMPTLESS
 	default y if !PSA_DEFAULT_OFF
-	depends on PSA_WANT_ALG_GCM
+	depends on PSA_WANT_ALG_GCM && !CRYPTOCELL_CC310_USABLE
```
Comment (mswarowsky) [inline nrf_security/src/drivers/nrf_cc3xx/Kconfig:145]: Why is it not `CRYPTOCELL_CC310_USABLE `, I would have understood it that way that we use the CC3XX config, if the cryptocell is usable

Comment (Vge0rge) [reply nrf_security/src/drivers/nrf_cc3xx/Kconfig:145]: This seems resolved to me

Comment (Vge0rge) [reply nrf_security/src/drivers/nrf_cc3xx/Kconfig:145]: Is it a typo? Was that intended to be !CRYPTOCELL_CC312_USABLE ?

Comment (frkv) [reply nrf_security/src/drivers/nrf_cc3xx/Kconfig:145]: We support GCM on CC312 only (which is the inverse of !CRYPTOCELL_CC310_USABLE). But the negative is unneccesary, so changing this to CRYPTOCELL_CC312_USABLE

File: nrf_security/cmake/psa_crypto_config.cmake [hunk 9]
```diff
@@ -75,6 +75,16 @@ kconfig_check_and_set_base_to_one(MBEDTLS_PSA_ACCEL_KEY_TYPE_CHACHA20)
 
 # Convert nrf_cc3xx driver configurations
 kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_CC3XX)
+kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_HAS_CIPHER_SUPPORT_CC3XX)
+kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_HAS_AEAD_SUPPORT_CC3XX)
+kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_HAS_MAC_SUPPORT_CC3XX)
+kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_HAS_HASH_SUPPORT_CC3XX)
+kconfig_check_and_set_base_to_one(PSA_CRYPTO_DRIVER_HAS_KEY_DERIVATION_SUPPORT_CC3XX)
```
[the remark this answers is not in the document]
Comment (frkv) [reply nrf_security/cmake/psa_crypto_config.cmake:83]: Not supported yet. Removing

[the remark this answers is not in the document]
Comment (frkv) [reply nrf_security/cmake/psa_crypto_config.cmake]: Fixed

File: nrf_security/src/drivers/nrf_cc3xx/Kconfig (hunk 8 above)
Comment (frkv) [reply nrf_security/src/drivers/nrf_cc3xx/Kconfig]: Fixed

File: nrf_security/cmake/psa_crypto_config.cmake (hunk 9 above)
[the remark this answers is not in the document]
Comment (frkv) [reply nrf_security/cmake/psa_crypto_config.cmake]: Fixed

**Review by Vge0rge — APPROVED** [review]

**Pull request merged**